quasardb/reader/ts_row.hpp

```diff
@@ -0,0 +1,250 @@
+    {
+        return _timestamp;
+    }
+
+protected:
+    qdb_local_table_t _local_table;
+    qdb_timespec_t _timestamp;
+};
+
+/**
+ * Our 'fast' row is a pure, lazy list-type row that uses a column's offset for
+ * constant-time access to the column values. E.g. row[0], row[1] which will return
+ * a (short-lived) ts_value.
+ */
+class ts_fast_row : public ts_row
+{
+public:
+    ts_fast_row()
+        : ts_row()
+    {}
+
+    ts_fast_row(qdb_local_table_t local_table, ts_columns_t const & columns)
+        : ts_row(local_table)
+        , _columns(columns)
+    {}
+
+    std::vector<py::object> copy() const
+    {
+        std::vector<py::object> res;
+        res.reserve(_columns.size());
+
+        for (auto index = 0; index < _columns.size(); ++index)
+        {
+            // By first casting the value, we create a copy of the concrete type
+            // (e.g. the int64 or the blob) rather than the reference into the local
+            // table.
+            res.push_back(py::cast(get_item(index), py::return_value_policy::move));
+        }
+
+        return res;
+    }
+
+    ts_value get_item(int64_t index) const
+    {
+        // TODO(leon): we construct this object every time; should be without any heap
+        // allocations,but can we do without this?
+        return ts_value(_local_table, index, _columns.at(index).type);
+    }
+
+    void set_item(int64_t /* index */, int64_t /* value */)
+    {
+        // not implemented
+    }
+
+    std::string repr() const
+    {
+        // This is inefficient because we first cast everything to Python objects, get the
+        // string representation and then cast back to c++ strings, but it's correct and
+        // effective, and it's used for debugging purposes only.
+        auto xs         = copy();
+        std::string kvs = std::accumulate(xs.cbegin(), xs.cend(), std::string(), [](std::string & acc, auto x) {
+            std::string s = py::str(x);
+            return acc.empty() ? s : acc + ", " + s;
+        });
+
+        return "[" + kvs + "]";
+    }
+
+private:
+    ts_columns_t _columns;
+};
+
+/**
+ * Our 'dict' row is a much slower, dict-based row type that provides very convenient
+ * access to the columns by their name. It does, however,  index each and every row's
+ * columns eagerly, and as such will exhibit much worse performance.
+ */
+class ts_dict_row : public ts_row
+{
+public:
+    ts_dict_row()
+        : ts_row()
+    {}
+
+    ts_dict_row(qdb_local_table_t local_table, ts_columns_t const & columns)
+        : ts_row(local_table)
+        , _indexed_columns(detail::index_columns(columns))
+    {}
+
+    ts_dict_row(const ts_dict_row & rhs)
+        : ts_row(rhs._local_table)
+        , _indexed_columns(rhs._indexed_columns)
+    {}
+
+    std::map<std::string, py::object> copy() const
+    {
+        typedef std::map<std::string, py::object> map_type;
+        map_type res;
+
+        for (auto c : _indexed_columns)
```

Comment:
```suggestion
        for (const auto & c : _indexed_columns)
```
Check your usage of `auto`.